lib/dialogs/settings.jsx

```diff
@@ -37,6 +38,74 @@ export const SettingsDialog = React.createClass({
     });
   },
 
+  onSignOutRequested() {
+    // Safety first! Check for any unsynced notes before signing out.
+    const { onSignOut, noteBucket } = this.props;
+    const { notes } = this.props.appState;
+    const { getVersion } = noteBucket;
+
+    if (
+      noteBucket.hasLocalChanges((error, hasChanges) => {
+        if (hasChanges) {
+          this.showUnsyncedWarning();
+          return;
+        }
+
+        // Also check persisted store for any notes with version 0
+        const noteHasSynced = note =>
+          new Promise((resolve, reject) =>
+            getVersion(note.id, (e, v) => (e || v === 0 ? reject() : resolve()))
+          );
+
+        Promise.race(notes.map(noteHasSynced)).then(
+          () => onSignOut(), // All good, sign out now!
+          () => this.showUnsyncedWarning() // Show a warning to the user
+        );
+      })
+    );
```

Comment:
what's going on here? is there a missing `if` statement body? something is wrong.

Reply:
Whoops, I had an `if` there that didn't need to be there. Callbacks are silly.

Reply:
:smile: looks better now